heartbeater.go

```diff
@@ -0,0 +1,167 @@
+package channel
+
+import (
+	"github.com/sirupsen/logrus"
+	"sync/atomic"
+	"time"
+)
+
+// HeartbeatCallback provide an interface that is notified when various heartbeat events take place
+type HeartbeatCallback interface {
+	HeartbeatTx(ts int64)
+	HeartbeatRx(ts int64)
+	HeartbeatRespTx(ts int64)
+	HeartbeatRespRx(ts int64)
+	CheckHeartBeat()
+}
+
+// ConfigureHeartbeat setups up heartbeats on the given channel. It assumes that an equivalent setup happens on the
+// other side of the channel.
+//
+// When possible, heartbeats will be sent on existing traffic. When a heartbeat is due to be sent, the next message sent
+// will include a heartbeat header. If no message is sent by the time the checker runs on checkInterval, a standalone
+// heartbeat message will be sent.
+//
+// Similarly, when a message with a heartbeat header is received, the next sent message will have a header set with
+// the heartbeat response. If no message is sent within a few milliseconds, a standalone heartbeat response will be
+// sent
+func ConfigureHeartbeat(binding Binding, heartbeatInterval time.Duration, checkInterval time.Duration, cb HeartbeatCallback) {
+	hb := &heartbeater{
+		ch:                  binding.GetChannel(),
+		heartBeatIntervalMs: heartbeatInterval.Milliseconds(),
+		callback:            cb,
+		events:              make(chan heartbeatEvent, 4),
+	}
+
+	binding.AddReceiveHandler(ContentTypeHeartbeat, hb)
+	binding.AddTransformHandler(hb)
+
+	go hb.pulse(checkInterval)
+}
+
+type heartbeater struct {
+	ch                   Channel
+	lastHeartbeatTx      int64
+	heartBeatIntervalMs  int64
+	callback             HeartbeatCallback
+	events               chan heartbeatEvent
+	unrespondedHeartbeat int64
+}
+
+func (self *heartbeater) HandleReceive(*Message, Channel) {
+	// ignore incoming heartbeat events, everything is handled by the transformer
+}
+
+func (self *heartbeater) Rx(m *Message, _ Channel) {
+	if val, found := m.GetUint64Header(HeartbeatHeader); found {
+		select {
+		case self.events <- heartbeatRxEvent(val):
```

Comment:
could these block?

Reply:
ah, I meant to put to defaults in those selects. Fixed